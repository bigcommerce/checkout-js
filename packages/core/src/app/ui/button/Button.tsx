import classNames from 'classnames';
import React, { ButtonHTMLAttributes, FunctionComponent } from 'react';

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
    isFullWidth?: boolean;
    isLoading?: boolean;
    size?: ButtonSize;
    testId?: string;
    variant?: ButtonVariant;
}

export enum ButtonVariant {
    Primary = 'primary',
    Secondary = 'secondary',
    Action = 'action',
}

export enum ButtonSize {
    Small = 'small',
    Tiny = 'tiny',
    Large = 'large',
}

function getClassName(
    props: Pick<ButtonProps, 'className' | 'isFullWidth' | 'isLoading' | 'size' | 'variant'>,
) {
    const { className, isFullWidth, isLoading, size, variant } = props;

    return classNames(
        'button',
        className,
        { 'button--primary': variant === ButtonVariant.Primary },
        { 'button--tertiary': variant === ButtonVariant.Secondary },
        { 'button--action': variant === ButtonVariant.Action },
        { 'button--small': size === ButtonSize.Small },
        { 'button--tiny': size === ButtonSize.Tiny },
        { 'button--large': size === ButtonSize.Large },
        { 'button--slab': isFullWidth },
        {
            'optimizedCheckout-buttonPrimary':
                variant === ButtonVariant.Primary || variant === ButtonVariant.Action,
        },
        { 'optimizedCheckout-buttonSecondary': variant === ButtonVariant.Secondary },
        { 'is-loading': isLoading },
    );
}

const Button: FunctionComponent<ButtonProps> = ({
    children,
    className,
    disabled,
    isFullWidth,
    isLoading,
    size,
    testId,
    type,
    variant,
    ...rest
}) => (
    <button
        {...rest}
        className={getClassName({ className, isFullWidth, isLoading, size, variant })}
        data-test={testId}
        disabled={disabled || isLoading}
        type={type || 'button'}
    >
        {children}
    </button>
);

export default Button;
