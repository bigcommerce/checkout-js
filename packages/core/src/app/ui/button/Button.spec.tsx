import { shallow } from 'enzyme';
import React from 'react';

import Button, { ButtonSize, ButtonVariant } from './Button';

describe('Button', () => {
    it('matches snapshot', () => {
        expect(shallow(<Button>Hello world</Button>))
            .toMatchSnapshot();
    });

    it('renders button with label', () => {
        expect(shallow(<Button>Hello world</Button>).text())
            .toEqual('Hello world');
    });

    it('renders button in various sizes', () => {
        expect(shallow(<Button size={ ButtonSize.Large }>Hello world</Button>).hasClass('button--large'))
            .toEqual(true);

        expect(shallow(<Button size={ ButtonSize.Small }>Hello world</Button>).hasClass('button--small'))
            .toEqual(true);

        expect(shallow(<Button size={ ButtonSize.Tiny }>Hello world</Button>).hasClass('button--tiny'))
            .toEqual(true);
    });

    it('renders button in various styles', () => {
        expect(shallow(<Button variant={ ButtonVariant.Primary }>Hello world</Button>).hasClass('button--primary'))
            .toEqual(true);

        expect(shallow(<Button variant={ ButtonVariant.Secondary }>Hello world</Button>).hasClass('button--tertiary'))
            .toEqual(true);

        expect(shallow(<Button variant={ ButtonVariant.Action }>Hello world</Button>).hasClass('button--action'))
            .toEqual(true);
    });

    it('renders button in full width', () => {
        expect(shallow(<Button isFullWidth>Hello world</Button>).hasClass('button--slab'))
            .toEqual(true);

        expect(shallow(<Button>Hello world</Button>).hasClass('button--slab'))
            .toEqual(false);
    });

    it('shows loading indicator', () => {
        expect(shallow(<Button isLoading>Hello world</Button>).hasClass('is-loading'))
            .toEqual(true);

        expect(shallow(<Button>Hello world</Button>).hasClass('is-loading'))
            .toEqual(false);
    });

    it('listens to DOM events', () => {
        const handleClick = jest.fn();
        const component = shallow(
            <Button onClick={ handleClick }>
                Hello world
            </Button>
        );

        component.simulate('click');

        expect(handleClick).toHaveBeenCalled();
    });
});
