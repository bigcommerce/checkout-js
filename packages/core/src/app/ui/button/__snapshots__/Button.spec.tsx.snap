// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button matches snapshot 1`] = `
<button
  className="button"
  type="button"
>
  Hello world
</button>
`;
